### PR TITLE
zcash_client_sqlite: Relax exclusions on wallet-internal ephemeral UTXOs

### PR DESCRIPTION
**Pull request by nuttycom**
In zcash/librustzcash#1985, a clause was added to UTXO queries to ensure that ephemeral outputs produced by wallet-internal transactions are excluded from no such outputs are ever regarded as spendable. However, this is too strict; it inhibits recovery of funds that end up "stuck" in an ephemeral address due to the partial failure of a TEX send. This relaxes that check such that if a TXO has actually been observed to be unspent after the second transaction in a TEX chain would have expired, we consider that unspentness information to be authoritative.

### CODE REVIEW

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 1]
```diff
@@ -995,21 +995,52 @@ pub(crate) fn tx_unexpired_condition_minconf_0(tx: &str) -> String {
...
             FROM v_received_output_spends
             WHERE v_received_output_spends.account_id = {accounts}.id
         )
+        -- or the transaction that generated the TXO would be considered expired as of the TXOs
+        -- max_observed_unspent_height; this operates under the assumption that the second
+        -- transaction in a TEX chain has the same expiry as the transaction that generated the
+        -- ephemeral output, and the output having been observed to be unspent above this height
+        -- indicates that the subsequent spend failed and the spending transaction will have
+        -- expired.
+        OR {transparent_received_outputs}.max_observed_unspent_height > {tx}.expiry_height
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet/transparent.rs:1043]: Bug: this should be
```suggestion
        OR {transparent_received_outputs}.max_observed_unspent_height >= {tx}.expiry_height
```
because the expiry height is the last valid height the tx can be mined at, so if the tx is observed unspent as of the end of the block at that height (which is what `max_observed_unspent_height` records), then the tx cannot be mined.

Add a test that catches this bug by exercising the boundary.

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 2]
```diff
@@ -1250,21 +1263,15 @@ pub(crate) fn get_transparent_balances<P: consensus::Parameters>(
...
              )
-             -- and the output is unspent
-             AND u.id NOT IN ({})
-             -- excluding received txos where the address is an ephemeral address and the
-             -- transaction that produced the TXO spent outputs belonging to the same account.
-             AND NOT (
-                addresses.key_scope = :ephemeral_key_scope
-                AND {} -- the transaction that created the txo spends account-internal outputs
-             )",
+
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet/transparent.rs]: Nit:
```suggestion
```

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed 3110d0a693ea8e04794198c1fa3acdc86f9db98c

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 1 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet/transparent.rs:1043]: This isn't a bug; using `>` instead of `>=` is the more conservative choice: we only allow the TXO to be selected once it has been observed to be unspent at a height _greater_ than the expiry height.

In fact, I think your suggested change here would be a bug. Note that the sense of this whole clause has changed, I rewrote it via De Morgan's law in order to be able to include this third clause via an `OR`.

Comment (str4d) [reply zcash_client_sqlite/src/wallet/transparent.rs:1043]: My brain was misparsing `unspent` as `unmined`, which led to my mis-analysis.

**Review by str4d — DISMISSED** [review]
utACK 3110d0a693ea8e04794198c1fa3acdc86f9db98c with prior nit.
